source/agora/common/Serializer.d

```diff
@@ -117,87 +110,72 @@ unittest
         dg(record[]);
-    else
-        foreach (const ref field; record.tupleof)
-            serializePart(field, dg);
-}
-
-/// Ditto
-public void serializePart (ubyte record, scope SerializeDg dg)
-    @trusted
-{
-    dg((cast(ubyte*)&record)[0 .. ubyte.sizeof]);
-}
-
-/// Ditto
-public void serializePart (ushort record, scope SerializeDg dg)
-    @trusted
-{
-    toVarInt(record, dg);
-}
 
-/// Ditto
-public void serializePart (uint record, scope SerializeDg dg)
-    @trusted
-{
-    toVarInt(record, dg);
-}
+    // Strings can be encoded as binary data
+    else static if (isNarrowString!T)
+    {
+        toVarInt(record.length, dg);
+        dg(cast(const(ubyte[]))record);
+    }
 
-/// Ditto
-public void serializePart (long record, scope SerializeDg dg)
-    @trusted
-{
-    dg((cast(ubyte*)&record)[0 .. long.sizeof]);
-}
+    // If it's binary data, just copy it
+    else static if (is(immutable(T) == immutable(ubyte[])))
+    {
+        toVarInt(record.length, dg);
+        dg(record);
+    }
 
-/// Ditto
-public void serializePart (ulong record, scope SerializeDg dg, CompactMode compact = CompactMode.Yes)
-    @trusted
-{
-    if (compact == CompactMode.Yes)
-        toVarInt(record, dg);
-    else
-        dg((cast(ubyte*)&record)[0 .. ulong.sizeof]);
-}
+    // If there's not fast path for array optimization, just recurse
+    else static if (is(T : E[], E))
+    {
+        toVarInt(record.length, dg);
+        foreach (const ref entry; record)
+            entry.serializePart(dg);
+    }
 
-/// Ditto
-public void serializePart (scope cstring record, scope SerializeDg dg)
-    @trusted
-{
-    serializePart(record.length, dg);
-    dg(cast(const ubyte[])record);
-}
+    // Unsigned integer may be encoded (e.g. sizes)
+    // However `ubyte` doesn't need binary encoding since they it already is the
+    // smallest possible size
+    else static if (is(Unqual!T == ubyte))
+        () @trusted { dg((&record)[0 .. T.sizeof]); }();
+    else static if (isUnsigned!T)
+    {
+        if (compact == CompactMode.Yes)
+            toVarInt(record, dg);
+        else
+            () @trusted { dg((cast(const(ubyte)*)&record)[0 .. T.sizeof]); }();
+    }
+    // Other integers / scalars
+    else static if (isScalarType!T)
+        () @trusted { dg((cast(const(ubyte)*)&record)[0 .. T.sizeof]); }();
 
-/// Ditto
-public void serializePart (scope const(ubyte)[] record, scope SerializeDg dg)
-    @safe
-{
-    serializePart(record.length, dg);
-    dg(record);
-}
+    // Recursively serialize fields for structs
+    else static if (is(T == struct))
+        foreach (const ref field; record.tupleof)
+            serializePart(field, dg);
 
-/// Ditto
-public void serializePart (scope const(Hash)[] record, scope SerializeDg dg)
-    @trusted
-{
-    serializePart(record.length, dg);
-    foreach (hash; record)
-        serializePart(hash, dg);
+    else
+        static assert(0, "Unhandled type: " ~ T.stringof);
```

Comment:
I'm wondering if it's right to cause an assert() error here.

Is this the point where you can attack an unacceptable type?

Reply:
The assert is static, which means it will trigger at compile time and you will know you need to fix it.

For example at the moment if you do:
```
    int[string] aa;
    serializeFull(aa);
```

You will see:
```console
agora ~serializer-improvements: building configuration "unittest"...
Running pre-build commands...
Target /Users/geod24/projects/bpfk/agora/source/scpp/build/libscp.o is up to date, nothing to do
source/agora/common/Serializer.d(181,9): Error: static assert:  "Unhandled type: const(int)[string]"
source/agora/common/Serializer.d(82,18):        instantiated from here: serializePart!(const(int)[string])
source/agora/common/Serializer.d(49,18):        instantiated from here: serializeFull!(int[string])
```

So you know the type can't be serialized and you need to add support to the serializer.

Reply:
Thank you for your detailed explanation.